[blob] implement client upload (#179)
vercel/storage: packages/blob/README.md

}> {}
```

### generateClientTokenFromReadWriteToken(options)

> Generates a single-use token which can be used from within the client.

Reviewer: ```suggestion
Generates a single-use token that can be used from within the client. This is useful when [uploading directly from browsers](#uploading-directly-from-browsers) to circumvent the 4MB limitation of going through a Vercel-hosted route.
```

Not sure about the markdown link 👍

Author: link works 👍

```ts
async function generateClientTokenFromReadWriteToken(options?: {
  token?: string;
  pathname?: string;
  onUploadCompleted?: {
    callbackUrl: string;
    metadata?: string;
  };
  maximumSizeInBytes?: number;
  allowedContentTypes?: string[];
}): string {}
```

> ## Examples

Reviewer: ```suggestion
Note: This method should be called server-side, not client-side.

## Examples
```

}> {}
```

### generateClientTokenFromReadWriteToken(options)

Generates a single-use token that can be used from within the client. This is useful when [uploading directly from browsers](#uploading-directly-from-browsers) to circumvent the 4MB limitation of going through a Vercel-hosted route.

> 

Reviewer: ```suggestion

Once created, a client token is valid for 30 seconds. This means you have 30 seconds to initiate an upload with this token.
```

      <form
        onSubmit={async (event): Promise<void> => {
          event.preventDefault();

          const file = inputFileRef.current?.files?.[0];
          if (!file) {
            return;
          }

          const clientTokenData = (await fetch('/api/upload-token', {
            method: 'POST',
            body: JSON.stringify({
              pathname: file.name,
            }),
          }).then((r) => r.json())) as { clientToken: string };

          const blobResult = await put(file.name, file, {
            access: 'public',
            token: clientTokenData.clientToken,
          });

          setBlob(blobResult);
        }}
      >
        <input name="file" ref={inputFileRef} type="file" />
        <button type="submit">Upload</button>
      </form>
      {blob && (
        <div>
          Blob url: <a href={blob.url}>{blob.url}</a>
        </div>
      )}
    </>
  );
}
```

```ts
// /app/api/upload-token/route.ts

import {
  generateClientTokenFromReadWriteToken,
  type GenerateClientTokenOptions,
} from '@vercel/blob';
import { NextResponse } from 'next/server';

export async function POST(request: Request): Promise<NextResponse> {

>   const body = (await request.json()) as GenerateClientTokenOptions;

Reviewer: I don't think we should reuse this type at this point.
Because metadata and callbackUrl can't be set by the browser.

We better just accept pathname

>   const body = (await request.json()) as GenerateClientTokenOptions;

Reviewer: ```suggestion
  // On a real website, this route would be protected by authentication, see: https://nextjs.org/docs/pages/building-your-application/routing/authenticating
  // Here, we accept the `pathname` from the browser, but in some situations, you may even craft the pathname
  // based on the authentication result
  const body = (await request.json()) as { pathname: string };
```

  return NextResponse.json({
    clientToken: await generateClientTokenFromReadWriteToken({
      ...body,
      onUploadCompleted: {
        callbackUrl: `https://${
          process.env.VERCEL_URL ?? ''
        }/api/upload-completed`,
        metadata: JSON.stringify({ foo: 'bar' }),
      },
    }),
  });
}
```

```ts
// /app/api/upload-completed/route.ts

import { type BlobUploadCompletedEvent } from '@vercel/blob';
import { NextResponse } from 'next/server';

export async function POST(request: Request): Promise<NextResponse> {

>   const body = (await request.json()) as BlobUploadCompletedEvent;

Reviewer: TODO: check signature

Author: https://linear.app/vercel/issue/STO-656/implement-webhook-signature-verification-helper

import { useState } from 'react';

export default function UploadForm() {
  const inputFileRef = useRef<HTMLInputElement>(null);
  const [blob, setBlob] = useState<BlobResult | null>(null);
  return (
    <>
      <h1>App Router Client Upload</h1>

      <form
        onSubmit={async (event): Promise<void> => {
          event.preventDefault();

          const file = inputFileRef.current?.files?.[0];
          if (!file) {
            return;
          }

          const clientTokenData = (await fetch('/api/upload-token', {
            method: 'POST',
            body: JSON.stringify({
              pathname: file.name,
            }),
          }).then((r) => r.json())) as { clientToken: string };

          const blobResult = await put(file.name, file, {
            access: 'public',
            token: clientTokenData.clientToken,
          });

          setBlob(blobResult);
        }}
      >
        <input name="file" ref={inputFileRef} type="file" />
        <button type="submit">Upload</button>
      </form>
      {blob && (
        <div>
          Blob url: <a href={blob.url}>{blob.url}</a>
        </div>
      )}
    </>
  );
}
```

```ts

> // /app/api/upload-token/route.ts

Reviewer: Should we be explicit and call this route "/app/api/generate-blob-client-token/route.ts"?

import {
  generateClientTokenFromReadWriteToken,
  type GenerateClientTokenOptions,
} from '@vercel/blob';
import { NextResponse } from 'next/server';

export async function POST(request: Request): Promise<NextResponse> {
  // On a real website, this route would be protected by authentication, see: https://nextjs.org/docs/pages/building-your-application/routing/authenticating
  // Here, we accept the `pathname` from the browser, but in some situations, you may even craft the pathname
  // based on the authentication result
  const body = (await request.json()) as { pathname: string };

  return NextResponse.json({
    clientToken: await generateClientTokenFromReadWriteToken({
      ...body,
      onUploadCompleted: {
        callbackUrl: `https://${
          process.env.VERCEL_URL ?? ''
        }/api/upload-completed`,

>         metadata: JSON.stringify({ foo: 'bar' }),

Reviewer: Let's avoid foo: 'bar' examples and put more real world usecases.

```suggestion
        metadata: JSON.stringify({ userId: 12345 }),
```

      },
      maximumSizeInBytes: 10_000_000, // 10 Mb
      allowedContentTypes: 'text/plain',
    }),
  });
}
```

```ts

> // /app/api/upload-completed/route.ts

Reviewer: Should we be explicit and call this route `/app/api/file-upload-completed/route.ts`?

> import { type BlobUploadCompletedEvent } from '@vercel/blob';

Reviewer: ```suggestion
import { type BlobUploadCompletedEvent, verifyCallbackSignature } from '@vercel/blob';
```

double check missing imports in the README if necessary 👍

import { NextResponse } from 'next/server';

export async function POST(request: Request): Promise<NextResponse> {
  const body = (await request.json()) as BlobUploadCompletedEvent;

>   console.log(body);

Reviewer: ```suggestion

```

Author: I would keep this line (and the next one showing the output)

Reviewer: I've added a suggestion that does that in a more explicit way, let me know

Reviewer: let's keep both 👍 let's just add one line break to breathe (updating suggestions)

  // { type: "blob.upload-completed", payload: { metadata: "{ foo: 'bar' }", blob: ... }}
  if (
    !(await verifyCallbackSignature({
      signature: request.headers.get('x-vercel-signature') ?? '',
      body: JSON.stringify(body),
    }))
  ) {
    return NextResponse.json(
      {
        response: 'invalid signature',
      },
      {
        status: 403,
      },
    );
  }

>   return NextResponse.json({

Reviewer: ```suggestion
  const metadata = JSON.stringify(body.payload.metadata);
  const blob = body.blob;
  
  console.log(metadata.userId); // 12345
  console.log(blob); // { url: '...', size: ..., uploadedAt: ..., ... }
  
  return NextResponse.json({
```

export async function POST(request: Request): Promise<NextResponse> {
  // On a real website, this route would be protected by authentication, see: https://nextjs.org/docs/pages/building-your-application/routing/authenticating
  // Here, we accept the `pathname` from the browser, but in some situations, you may even craft the pathname
  // based on the authentication result
  const body = (await request.json()) as { pathname: string };

  return NextResponse.json({
    clientToken: await generateClientTokenFromReadWriteToken({
      ...body,
      onUploadCompleted: {
        callbackUrl: `https://${
          process.env.VERCEL_URL ?? ''
        }/api/upload-completed`,
        metadata: JSON.stringify({ foo: 'bar' }),
      },
      maximumSizeInBytes: 10_000_000, // 10 Mb
      allowedContentTypes: 'text/plain',
    }),
  });
}
```

```ts
// /app/api/upload-completed/route.ts

import { type BlobUploadCompletedEvent, verifyCallbackSignature } from '@vercel/blob';
import { NextResponse } from 'next/server';

export async function POST(request: Request): Promise<NextResponse> {
  const body = (await request.json()) as BlobUploadCompletedEvent;
  console.log(body);

>   // { type: "blob.upload-completed", payload: { metadata: "{ foo: 'bar' }", blob: ... }}

Reviewer: ```suggestion
  // { type: "blob.upload-completed", payload: { metadata: "{ foo: 'bar' }", blob: ... }}
  
```

Adding a breather line
